### PR TITLE
feat: add data processing notebooks

### PR DESCRIPTION
**Pull request by SaFouNaldo**
## Project layout :
project_root/
│
├─ penn_action_raw/
│ └─ Penn_Action/ ← official dataset (replace this folder after extraction)
│
├─ notebooks/ ← the six notebooks live here
│
├─ output/ ← artifacts produced by the notebooks
│ ├─ train/ ← full-dataset token files
│ ├─ dump/ ← 1-video demo dump
> **Note:** All notebooks assume the working directory is the project root,  
> so relative paths like `penn_action_raw/Penn_Action/frames/...` resolve correctly.

---

## Notebooks at a glance

| # | Notebook | Purpose |
|---|----------|---------|
| 1 | **`data_set_processing.ipynb`** | Pre-processing & sanity checks |
| 2 | **`data_tokenize_all.ipynb`** | Tokenise **entire** dataset (caption, pose, RGB, coords) |
| 3 | **`data_tokenize_all_example.ipynb`** | One-video smoke test of the full pipeline |
| 4 | **`data_tokenize_caption.ipynb`** | Explore/visualise caption tokenisation schemes |
| 5 | **`data_tokenize_human_poses.ipynb`** | Pose tokenisation & rendering analysis |
| 6 | **`data_tokenize_rgb.ipynb`** | RGB frame tokenisation & reconstruction visuals | 

---

## Suggested run order

1. `data_set_processing.ipynb`
2. `data_tokenize_rgb.ipynb`
3. `data_tokenize_human_poses.ipynb`
4. `data_tokenize_caption.ipynb`
5. `data_tokenize_all_example.ipynb`
6. `data_tokenize_all.ipynb`

---
## Next steps: 
- Train the model using the run_training in the nano4M
--- 
## Useful tips:
- To download the data set: 
1.  go to [this](https://www.kaggle.com/datasets/kaushalbora18/penn-action-dataset)
2. Create an account and go to settings > API > Create New API key this downloads kaggle.json.
3. In your local machine run the following command more info [here](https://scitas-doc.epfl.ch/user-guide/data-management/transferring-data/): `$ rsync -azP  ~/Downloads/kaggle.json <GASPAR>@<machine>.hpc.epfl.ch:/path/to/dest`
4.  On izar terminal rund the follwoing command: 
- `pip install --quiet --upgrade pip kaggle`
- `kaggle datasets download -d kaushalbora18/penn-action-dataset --unzip \-p ~/path/to/dest`

### CODE REVIEW

**Comment by SaFouNaldo** [comment]
I've added 4 files:
- `multi_penn_action.yaml`: The configuration file for training.

- `create_penn_action_masked_dataloader.py`: Script for instantiating the data loader.

- `penn_action_multimodal_dataset.py`: The new data loader for Penn Action.

- `data_penn_view.ipynb`: A Jupyter notebook to visualize the data & masked data in `work/com-304/CFT/output/train`.

**Comment by SaFouNaldo** [comment]
# Overview:
- This PR brings together all the infrastructure and fixes we developed while stabilizing Penn-Action training. In particular:

  1. New Transformer Layers

  2. `run_training.py` Enhancements

  3. FourM: Three Watchdogs for NaN Protection

  4. Updated Default Config


### 1. Transformer Layers
- Added TransformerTrunk and TransformerDecoderTrunk same in `part3 `of the notebooks

### 2. `run_training.py` Enhancements:
- Debug flags and assertion hooks in the very first batch, to catch:

     - Out-of-range token IDs

     - NaN/Inf activations in forward/backward

- What they solved:

     - Pinpointed two root causes: an inverted pad mask in the decoder, and occasional all--inf softmax rows.

     - Let us fail immediately with clear error messages instead of hours of silent NaNs.
**Most Important changes**  `Robust Decoder Masking in forward_decoder` :
- Replaced the simple `repeat(~mask) `logic with a four‐step recipe:
- 
```
# 1) derive “valid” boolean masks (True = keep)
valid_q   = (dec_pad_mask if dec_pad_mask is not None
             else torch.ones(B, M, device=x.device)).bool()
valid_enc = (enc_pad_mask if enc_pad_mask is not None
             else torch.ones(B, N, device=x.device)).bool()

# 2) build full self- and cross-attention masks
dec_pad_sa_mask = valid_q.unsqueeze(2) & valid_q.unsqueeze(1)   # (B, M, M)
dec_pad_xa_mask = valid_q.unsqueeze(2) & valid_enc.unsqueeze(1) # (B, M, N)

# 3) ensure no query row is all False (which would give softmax(–inf)→NaN)
row_ok_sa = dec_pad_sa_mask.any(-1, keepdim=True)
row_ok_xa = dec_pad_xa_mask.any(-1, keepdim=True)
dec_pad_sa_mask = torch.where(row_ok_sa, dec_pad_sa_mask,
                              torch.ones_like(dec_pad_sa_mask))
dec_pad_xa_mask = torch.where(row_ok_xa, dec_pad_xa_mask,
                              torch.ones_like(dec_pad_xa_mask))

# 4) zero-out embeddings of padding queries so they can’t pollute 
x = x.masked_fill(~valid_q.unsqueeze(-1), 0.)
```
- `Problem solved`: eliminated the “entirely masked query → all –inf → NaN” failure in the decoder.

- `Why it matters`: Without this, padded positions with no valid keys would crash training unpredictably.

### 3. FourM: Three Watchdogs for NaN Protection:
- `In nanofm/models/fourm.py` we added three “watchdog” checks in compute_ce_loss (in a simplified version):
- **Logits Finite Check**: `assert torch.isfinite(logits).all(), "non-finite logits"` 
- **Target‐ID Range Check**: `assert ((target_seq == pad_idx) | (0 <= target_seq < vocab_size)).all(), "bad target id"`
- **Parameter Finite Scan**:  
```
for n, p in self.named_parameters(): …
     assert torch.isfinite(p).all(), f"bad param {n}"
```
- Problem addressed: silent NaNs in LogSoftmaxBackward0 and impossible-to-debug loss spikes.

- Where: in `compute_ce_loss.`

- Result: any numerical instability now halts with a clear message indicating exactly which tensor or sample blew up.

### 4. Config Changes:
- Switched default precision to **fp32** on V100s (safer than fp16 for this model).

- Increased decoder depth from 1 layer → 8 layers to better handle longer sequences.

- Vocab size to [65536,65536,65536] with overlapping pose emb

**Review by GhaliBerbich — APPROVED** [review]
